src/sqlancer/StatementReducer.java

```diff
@@ -1,59 +1,95 @@
 package sqlancer;
 
+import java.time.Duration;
+import java.time.Instant;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.function.BiFunction;
-import java.util.stream.Collectors;
 
 import sqlancer.common.query.Query;
 
 public class StatementReducer<G extends GlobalState<O, ?, C>, O extends DBMSSpecificOptions<?>, C extends SQLancerDBConnection>
         implements Reducer<G> {
     private final DatabaseProvider<G, O, C> provider;
     private boolean observedChange;
+    private int partitionNum;
 
     public StatementReducer(DatabaseProvider<G, O, C> provider) {
         this.provider = provider;
     }
 
+    private boolean hasNotReachedLimit(long curr, long limit) {
+        if (limit > 0) {
+            return curr < limit;
+        }
+        return true;
```

Comment:
I think this would be easier to read if you first check for the special value `-1` and return `true` then. You could also define a constant in `MainOptions` for that, so that future changes keep in sync.